File: bot/document_processor/factory.py
from bot.utils.logging import logger
from bot.document_processor.base import DocumentProccesor
from bot.document_processor.direct_link import DirectLink
from bot.document_processor.gdrive import GDrive


class DocumentProcessorFactory:

    @staticmethod
    def create_document_processor(url: str, message) -> DocumentProccesor:
        """The method which decides the concrete implementation
        to be used to process a particular file.
        This decision is made based on the url being provided.
        Parameters
            url: URL of the file to be downloaded.
        Returns
            object: The object of the concrete Document Processor class to be used
        """
        if "drive.google.com" in url:
            logger.info("The file is hosted on Google Drive")
            return GDrive(message)
        logger.info("The file is a direct link download")
        return DirectLink(message)

File: src/document_processor/factory.rs
use grammers_client::{types::Message, Client};

use super::{direct_link::DirectLink, document::Document, gdrive::Gdrive};

pub enum DocumentProcessor<S, T> {
    GDrive(S),
    DirectLink(T),
}
pub fn factory(
    link: String,
    client: Client,
    message: Message,
) -> DocumentProcessor<Gdrive, DirectLink> {
    if link.contains("drive.google.com") {
        DocumentProcessor::GDrive(Gdrive::new(Document::new(
            message,
            client,
            "gdrive".to_string(),
        )))
    } else {
        DocumentProcessor::DirectLink(DirectLink::new(Document::new(
            message,
            client,
            "direct_link".to_string(),
        )))
    }
}
